migrations/versions/20181212_rename_page_table.py

```diff
@@ -0,0 +1,160 @@
+"""Rename page table to `measure_version` and other changes cascading from that
+
+This migration will require publisher downtime as it will lock pretty much the entire database - and is not backwards
+compatible.
+
+Revision ID: 20181212_rename_page_table
+Revises: 2018_12_05_not_nullable_fields
+Create Date: 2018-12-03 10:51:52.822365
+
+"""
+from alembic import op
+import sqlalchemy as sa
+from sqlalchemy.schema import Sequence, CreateSequence, DropSequence
+
+
+# revision identifiers, used by Alembic.
+revision = "20181212_rename_page_table"
+down_revision = "2018_12_05_not_nullable_fields"
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # * remove constraints
+    op.drop_constraint("data_source_in_page_page_guid_fkey", "data_source_in_page", type_="foreignkey")
+    op.drop_constraint("dimension_page_id_version_fkey", "dimension", type_="foreignkey")
+    op.drop_constraint("upload_page_id_version_fkey", "upload", type_="foreignkey")
+    op.drop_constraint("page_parent_guid_version_fkey", "page", type_="foreignkey")
+    op.drop_constraint("uq_page_guid_version", "page", type_="unique")
+    op.drop_constraint("page_guid_version_pkey", "page", type_="primary")
+    op.drop_constraint("data_source_id_page_guid_version_pk", "data_source_in_page", type_="primary")
+
+    # * create `measure_version_id_seq`
+    op.execute(CreateSequence(Sequence("measure_version_id_seq")))
+
+    # * add columns
+    #   * `id` to `page`, not nullable, default next from `measure_version_id_seq`
+    op.add_column(
+        "page",
+        sa.Column(
+            "id", sa.Integer(), nullable=False, server_default=sa.text("nextval('measure_version_id_seq'::regclass)")
+        ),
+    )
+    op.add_column("page", sa.Column("parent_id", sa.Integer(), nullable=True))
+    op.add_column("data_source_in_page", sa.Column("measure_version_id", sa.Integer(), nullable=True))
+    op.add_column("upload", sa.Column("measure_version_id", sa.Integer(), nullable=True))
+    op.add_column("dimension", sa.Column("measure_version_id", sa.Integer(), nullable=True))
+
+    conn = op.get_bind()
+
+    conn.execute(
+        "UPDATE page AS child SET parent_id = parent.id FROM page AS parent WHERE child.parent_guid = parent.guid AND child.parent_version = parent.version AND child.page_type = 'measure' and parent.page_type = 'measure'"
+    )
+    conn.execute(
+        "UPDATE upload SET measure_version_id = page.id FROM page WHERE upload.page_id = page.guid AND upload.page_version = page.version"
+    )
+    conn.execute(
+        "UPDATE dimension SET measure_version_id = page.id FROM page WHERE dimension.page_id = page.guid AND dimension.page_version = page.version"
+    )
+    conn.execute(
+        "UPDATE data_source_in_page SET measure_version_id = page.id FROM page WHERE data_source_in_page.page_guid = page.guid AND data_source_in_page.page_version = page.version"
+    )
+
+    op.alter_column("page", "id", nullable=False)
```

Comment:
The `id` field already has `nullable=False` when it was created so no need to re-do it here.

Reply:
👍